jenkins/control-slave.sh

```diff
@@ -0,0 +1,146 @@
+IMAGE='debian-7-backports'
+MACHINE_TYPE='n1-highmem-32'  # TODO(fejta): optimize to lower values
+SCOPES='cloud-platform,compute-rw,storage-full'  # TODO(fejta): verify
+
+if [[ -z "${1:-}" ]]; then
+  echo "Usage: $(basename "${0}") --small <instance> [AUTO|create|delete|update|reboot|attach]"
+  exit 1
+fi
+
+if [[ "${1:-}" == "--small" ]]; then
+  shift
+  echo 'Reducing resource footprint for script testing...'
+  DISK_SIZE='200GB'
+  DISK_TYPE='pd-standard'
+  MACHINE_TYPE='n1-standard-1'
+fi
+
+INSTANCE="${1}"
+CMD="${2:-auto}"
+
+auto-slave() {
+  echo "Automatically creating ${INSTANCE}..."
+  delete-slave
+  create-slave
+  update-slave
+  reboot-slave
+  attach-slave
+}
+
+attach-slave() {
+  gcloud compute ssh "${MASTER}" --ssh-flag='-L8000:localhost:8080' sleep 60 &
+  echo "not done yet"
+  wait
+}
+
+delete-slave() {
+  echo "Delete ${INSTANCE}..."
+  if [[ -z "$(gcloud compute instances list "--filter=name=${INSTANCE}")" ]]; then
+    return 0
+  fi
+  gcloud -q compute instances delete "${INSTANCE}"
+}
+
+create-slave() {
+  echo "Create ${INSTANCE}..."
+  gcloud compute instances create \
+    "${INSTANCE}" \
+    "--boot-disk-size=${DISK_SIZE}" \
+    "--boot-disk-type=${DISK_TYPE}" \
+    "--image=${IMAGE}" \
+    "--machine-type=${MACHINE_TYPE}" \
+    "--scopes=${SCOPES}"
+}
+
+update-slave() {
+
+echo "Instantiate ${INSTANCE}..."
+gcloud compute ssh "${INSTANCE}" << INSTANTIATE_DONE
+set -o verbose
+set -o errexit
+
+# Install docker
+which docker || curl -sSL https://get.docker.com/ | sh
+id jenkins || sudo useradd jenkins -m
+sudo usermod -aG docker jenkins
+
+# Downgrade to 1.9.1
+# https://github.com/kubernetes/kubernetes/issues/21451
+sudo apt-get -y --force-yes install docker-engine="${DOCKER_VERSION}"
+sudo docker run hello-world
+
+# Install go (needed for hack/e2e.go)
+
+wget "https://storage.googleapis.com/golang/${GO_VERSION}"
+sudo tar xzvf "${GO_VERSION}" -C /usr/local
+
+# install build tools
+sudo apt-get -y install build-essential
+
+# Install python
+sudo apt-get -y install python-openssl python-pyasn1 python-ndg-httpsclient
+
+# Reboot on panic
+sudo touch /etc/sysctl.conf
+sudo sh -c 'cat << END >> /etc/sysctl.conf
+kernel.panic_on_oops = 1
+kernel.panic = 10
+END'
+sudo sysctl -p
+
+# Keep tmp clean
+sudo apt-get -y install tmpreaper
+sudo sh -c "cat << END > /etc/tmpreaper.conf
+TMPREAPER_PROTECT_EXTRA=''
+TMPREAPER_DIRS='/tmp/. /var/tmp/.'
+TMPREAPER_TIME='3'
+TMPREAPER_DELAY='256'
+TMPREAPER_ADDITIONALOPTIONS=''
+END"
+
+# Configure the time zone
+sudo sh -c 'echo ${TIMEZONE} > /etc/timezone'
+sudo dpkg-reconfigure -f noninteractive tzdata
+
+# Prepare jenkins workspace
+sudo mkdir -p /var/lib/jenkins
+sudo chown jenkins:jenkins /var/lib/jenkins
+sudo su jenkins
+set -o verbose
+set -o errexit
+
+gcloud compute config-ssh
+mkdir -p /var/lib/jenkins/gce_keys
+chown jenkins:jenkins /var/lib/jenkins/gce_keys
+cp ~/.ssh/google_compute_engine* /var/lib/jenkins/gce_keys/
+exit
```

Comment:
Something like..

```
sudo --user jenkins <<EOF
set -o ...
....
EOF
```

might be a little easier to read, though it has its own downsides of it being code in a heredoc. Either approach seems fine, but with this one I think a comment on `exit` is warranted because this pattern isn't too common.